readthedocs/doc_builder/python_environments.py

```diff
@@ -72,6 +74,103 @@ def delete_existing_venv_dir(self):
             )
             shutil.rmtree(venv_dir)
 
+    def install_languages(self):
+        if settings.RTD_DOCKER_COMPOSE:
+            # Create a symlink for ``root`` user to use the same ``.asdf``
+            # installation than ``docs`` user. Required for local building
+            # since everything is run as ``root`` when using Local Development
+            # instance
+            cmd = [
+                'ln',
+                '-s',
+                '/home/docs/.asdf',
+                '/root/.asdf',
+            ]
+            self.build_env.run(
+                *cmd,
+            )
+
+        # TODO: do not use a Feature flag here, but check for ``build.os`` and
+        # ``build.languages`` instead
+        if self.project.has_feature(Feature.USE_NEW_DOCKER_IMAGES_STRUCTURE):
+            # TODO: iterate over ``build.languages`` and install all languages
+            # specified for this project
+            build_os = 'ubuntu20'
+            language = 'python'
+            version = '3.9.6'
+
+            # TODO: generate the correct path for the Python version
+            # language_path = f'{build_os}/{language}/2021-08-30/{version}.tar.gz'
+            language_path = f'{build_os}-{language}-{version}.tar.gz'
+            language_version_cached = build_languages_storage.exists(language_path)
+            if language_version_cached:
+                remote_fd = build_languages_storage.open(language_path, mode='rb')
+                with tarfile.open(fileobj=remote_fd) as tar:
+                    # Extract it on the shared path between host and Docker container
+                    extract_path = os.path.join(self.project.doc_path, 'languages')
+                    tar.extractall(extract_path)
+
+                    # Move the extracted content to the ``asdf`` installation
+                    cmd = [
+                        'mv',
+                        f'{extract_path}/{version}',
+                        f'/home/docs/.asdf/installs/{language}/{version}',
+                    ]
+                    self.build_env.run(
+                        *cmd,
+                    )
+            else:
+                # If the language version selected is not available from the
+                # cache we compile it at build time
+                cmd = [
+                    # TODO: make this environment variable to work
+                    # 'PYTHON_CONFIGURE_OPTS="--enable-shared"',
+                    'asdf',
+                    'install',
+                    language,
+                    version,
+                ]
```

Comment:
We should definitely test this one more with wider sets of data. I could see this falling over because we compiled a Python without some necessary flag/etc.

Reply:
Currently, we are using exactly the same flags that we are using in our current production images (`asdf` uses `pyenv` `python-build` behind the scenes) so I'm not worried about this.

However, I wanted to introduce `--enabled-shared` here because it was asked by a user in https://github.com/readthedocs/readthedocs-docker-images/pull/130